montblanc/impl/rime/tensorflow/rime_ops/tests/test_zernike.py

```diff
@@ -0,0 +1,773 @@
+    _impl_test_zernike(FT, CT, gpu_devs, coeff_yx, noll_index_yx, 8, eidos_data_yx, 2)
+
+@pytest.mark.parametrize("FT, CT", [(np.float32, np.complex64), (np.float64, np.complex128)])
+def _test_zernike_yy(gpu_devs, coeff_yy, noll_index_yy, eidos_data_yy, FT, CT):
+    """ Test the Zernike operator """
+    _impl_test_zernike(FT, CT, gpu_devs, coeff_yy, noll_index_yy, 15, eidos_data_yy, 3)
+
+
+def _impl_test_zernike(FT, CT, gpu_devs, coeff_nn, noll_index_nn, thresh, eidos_data_nn, corr_num):
+    """ Implementation of the Zernike operator test """
+    npix = 17
+    nsrc = npix ** 2
+    ntime = 20
+    na = 20
+    nchan = 20
+
+    nx, ny = npix, npix
+    grid = (np.indices((nx, ny), dtype=np.float) - nx//2) * 2 / nx
+    ll, mm = grid[0], grid[1]
+
+    lm = np.vstack((ll.flatten(), mm.flatten())).T
+    # Create input variables
+    coords = np.empty((nsrc, 4, 2)).astype(FT)
+    coeffs = np.empty((na, nchan, thresh, 4)).astype(CT)
+    noll_index = np.zeros((na, nchan, thresh, 4)).astype(np.int32)
+    pointing_error = np.zeros((ntime, na, nchan, 4, 2)).astype(FT)
+    antenna_scaling = np.empty((na, nchan, 4, 2)).astype(FT)
+
+    antenna_scaling[:,:,:,:] = 1
+
+    coeffs[0,0,:, 0], coeffs[0,0,:, 1], coeffs[0,0,:, 2], coeffs[0,0,:, 3] = coeff_nn[:thresh], coeff_nn[:thresh], coeff_nn[:thresh], coeff_nn[:thresh]
+    noll_index[0,0,:, 0], noll_index[0,0,:, 1], noll_index[0,0,:, 2], noll_index[0,0,:, 3] = noll_index_nn[:thresh], noll_index_nn[:thresh], noll_index_nn[:thresh], noll_index_nn[:thresh]
+
+    coords[0:nsrc, 0, 0], coords[0:nsrc, 1, 0], coords[0:nsrc, 2, 0], coords[0:nsrc, 3, 0] = lm[0:nsrc, 0], lm[0:nsrc, 0], lm[0:nsrc, 0], lm[0:nsrc, 0]
+    coords[0:nsrc, 0, 1], coords[0:nsrc, 1, 1], coords[0:nsrc, 2, 1], coords[0:nsrc, 3, 1] = lm[0:nsrc, 1], lm[0:nsrc, 1], lm[0:nsrc, 1], lm[0:nsrc, 1]
+
+
+    # Argument list
+    np_args = [coords, coeffs, noll_index, pointing_error, antenna_scaling]
+    # Argument string name list
+    arg_names = ['coords', 'coeffs', 'noll_index', 'pointing_error', 'antenna_scaling']
+    # Constructor tensorflow variables
+    tf_args = [tf.Variable(v, name=n) for v, n in zip(np_args, arg_names)]
+
+    def _pin_op(device, *tf_args):
+        """ Pin operation to device """
+        with tf.device(device):
+            return zernike(*tf_args)
+
+    # Pin operation to CPU
+    cpu_op = _pin_op('/cpu:0', *tf_args)
+
+    # Run the op on all GPUs
+    gpu_ops = [_pin_op(d, *tf_args) for d in gpu_devs]
+
+    # Initialise variables
+    init_op = tf.global_variables_initializer()
+    with tf.Session() as S:
+        S.run(init_op)
+        cpu_data = S.run(cpu_op)
+        cpu_data = cpu_data[:, 0, 0, 0, corr_num].reshape((npix, npix))
+        assert np.allclose(cpu_data, eidos_data_nn, atol=1e-6, rtol=1e-4)
+        gpu_data = np.array(S.run(gpu_ops))
+        gpu_data = gpu_data[ 0, :, 0, 0, 0, corr_num].reshape((npix,npix))
+        assert np.allclose(gpu_data, eidos_data_nn, atol=1e-6, rtol=1e-4)
```

Comment:
I think that in addition to the parametrization on `FT` and `CT`, these tests should be parametrized on an `rtol` which can be relaxed for float32, but should be full precision for float64.